File: Kv/test_cli/kv_cli.cc
// MIT License

// Copyright (c) 2021 Colin

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <iostream>
#include <memory>
#include <string>

#include <grpcpp/grpcpp.h>
#include <eraftio/helloworld.grpc.pb.h>
#include <eraftio/kvrpcpb.grpc.pb.h>
#include <Kv/raft_client.h>
#include <Kv/config.h>

using namespace kvserver;

int main(int argc, char** argv) {

  std::shared_ptr<Config> conf = std::make_shared<Config>(std::string(argv[1]), 
    std::string(argv[2]), std::stoi(std::string(argv[3])));
  
  std::shared_ptr<RaftClient> raftClient = std::make_shared<RaftClient>(conf);
  kvrpcpb::RawPutRequest request;
  request.mutable_context()->set_region_id(1);
  request.set_cf(std::string(argv[2]));
  request.set_value(std::string(argv[2]));
  request.set_key(std::string(argv[2]));
  raftClient->PutRaw(std::string(argv[1]), request);

  return 0;
}

File: cmd/kvcli/kvcli.go
//
// MIT License

// Copyright (c) 2022 eraft dev group

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

package main

import (
	"context"
	"crypto/rand"
	"fmt"
	"math/big"
	"os"
	"os/signal"

	"github.com/eraft-io/eraft/common"
	pb "github.com/eraft-io/eraft/raftpb"

	"github.com/eraft-io/eraft/raftcore"
)

type KvClient struct {
	rpcCli    *raftcore.RaftClientEnd
	leaderId  int64
	clientId  int64
	commandId int64
}

func (kvCli *KvClient) Close() {
	kvCli.rpcCli.CloseAllConn()
}

func nrand() int64 {
	max := big.NewInt(int64(1) << 62)
	bigx, _ := rand.Int(rand.Reader, max)
	return bigx.Int64()
}

func MakeKvClient(targetId int, targetAddr string) *KvClient {
	cli := raftcore.MakeRaftClientEnd(targetAddr, uint64(targetId))
	return &KvClient{
		rpcCli:    cli,
		leaderId:  0,
		clientId:  nrand(),
		commandId: 0,
	}
}

func (kvCli *KvClient) Get(key string) string {
	cmdReq := &pb.CommandRequest{
		Key:      key,
		OpType:   pb.OpType_OpGet,
		ClientId: kvCli.clientId,
	}
	resp, err := (*kvCli.rpcCli.GetRaftServiceCli()).DoCommand(context.Background(), cmdReq)
	if err != nil {
		return "err"
	}
	return resp.Value
}

func (kvCli *KvClient) Put(key, value string) string {
	cmdReq := &pb.CommandRequest{
		Key:      key,
		Value:    value,
		ClientId: kvCli.clientId,
		OpType:   pb.OpType_OpPut,
	}
	_, err := (*kvCli.rpcCli.GetRaftServiceCli()).DoCommand(context.Background(), cmdReq)
	if err != nil {
		return "err"
	}
	return "ok"
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("usage: kvcli [serveraddr]")
		return
	}
	sigs := make(chan os.Signal, 1)

	kvCli := MakeKvClient(99, os.Args[1])

	sigChan := make(chan os.Signal)
	signal.Notify(sigChan)

	go func() {
		sig := <-sigs
		fmt.Println(sig)
		kvCli.rpcCli.CloseAllConn()
		os.Exit(-1)
	}()

	for i := 0; i <= 10000; i++ {
		fmt.Println(kvCli.Put(common.RandStringRunes(1024), common.RandStringRunes(256)))
	}

	// fmt.Println("run test get value -> " + kvCli.Get("testkey"))
}
